### PR TITLE
Watch for changes in types roots

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Watch for changes in type roots and reload the project when that happens.

Only tested in VS code so far

### CODE REVIEW

File: src/server/utilities.ts [hunk 1]
```diff
@@ -90,6 +90,21 @@ namespace ts.server {
...
     }
 
+    export function getLatestDirectoryChangeTime(paths: string[] | undefined, host: System) {
+        if (!host.getModifiedTime || !host.directoryExists || !paths) {
+            return 0;
+        }
+
+        return Math.max.apply(Math, paths.map(path => {
+            if (host.directoryExists(path)) {
+                return host.getModifiedTime(path);
```
Comment (billti) [inline src/server/utilities.ts]: This returns a Date, but this function should return a number (Math.max probably converts it implicitly here).

**Comment by billti** [comment]
👍

File: src/server/project.ts [hunk 2]
```diff
@@ -608,6 +610,18 @@ namespace ts.server {
             this.projectFileWatcher = this.projectService.host.watchFile(this.configFileName, _ => callback(this));
         }
 
+        watchTypeRoots(callback: (project: ConfiguredProject, path: string) => void) {
+            const roots = ts.getEffectiveTypeRoots(this.getCompilerOptions(), this.projectService.host);
```
Comment (mhegazy) [inline src/server/project.ts]: `this.getEffectiveTypeRoots()` ?

File: src/server/project.ts [hunk 3]
```diff
@@ -608,6 +610,18 @@ namespace ts.server {
             this.projectFileWatcher = this.projectService.host.watchFile(this.configFileName, _ => callback(this));
         }
 
+        watchTypeRoots(callback: (project: ConfiguredProject, path: string) => void) {
+            const roots = ts.getEffectiveTypeRoots(this.getCompilerOptions(), this.projectService.host);
+            const watchers: FileWatcher[] = [];
+            if (roots) {
+                for (const root of roots) {
+                    this.projectService.logger.info(`Add type root watcher for: ${root}`);
+                    watchers.push(this.projectService.host.watchDirectory(root, path => callback(this, path), true));
```
Comment (mhegazy) [inline src/server/project.ts]: why recursive? we only care about the top level.

File: src/server/editorServices.ts [hunk 4]
```diff
@@ -363,6 +363,11 @@ namespace ts.server {
             this.printProjects();
         }
 
+        private onTypeRootFileChanged(project: ConfiguredProject, fileName: string) {
+            this.updateConfiguredProject(project);
```
Comment (mhegazy) [inline src/server/editorServices.ts]: consider adding a log message here.

Comment (mhegazy) [inline src/server/editorServices.ts]: why not `this.throttledOperations.schedule` ? in case there is a flurry of these, e.g. `npm install`

**Comment by mhegazy** [comment]
@vladima can you take a look.

**Pull request merged**